tasks/ldap/src/main/java/com/walmartlabs/concord/plugins/ldap/CustomSocketFactory.java

```diff
@@ -0,0 +1,191 @@
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =====
+ */
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.net.SocketFactory;
+import javax.net.ssl.*;
+import java.io.BufferedInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.InetAddress;
+import java.net.Socket;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.KeyStore;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+
+public class CustomSocketFactory extends SocketFactory {
+    private static final Logger log = LoggerFactory.getLogger(CustomSocketFactory.class);
+    // reuse single instance via getDefault() method
+    private static CustomSocketFactory instance;
+    static final String CERT_PATH = ".ldapcerts";
+
+    private final SocketFactory sslSocketFactory;
+
+
+    public CustomSocketFactory() {
+        sslSocketFactory = createSSLSocketFactory();
+    }
+
+    public static SocketFactory getDefault() {
+        synchronized (CustomSocketFactory.class) {
+            if (instance == null) {
+                instance = new CustomSocketFactory();
+            }
+        }
+
+        return instance;
+    }
+
+    private static SocketFactory createSSLSocketFactory() {
+        try {
+            KeyStore keyStore = generateKeystore(getCerts());
+            X509TrustManager tm = getTrustManager(keyStore);
+            SSLContext sslContext = getSslContext(keyStore, tm);
+
+            return sslContext.getSocketFactory();
+        } catch (Exception e) {
+            log.error("Error setting up custom socket factory", e);
+        }
+
+        throw new RuntimeException("Something went wrong generating socket factory");
+    }
+
+    /**
+     * Creates an SSL Context for validating SSL connections.
+     *
+     * @param keyStore KeyStore to use for certificate validations
+     * @param tm       for making trust decisions
+     * @return SSL Context with the given keystore and trustmanager
+     * @throws Exception
+     */
+    private static SSLContext getSslContext(KeyStore keyStore, X509TrustManager tm) throws Exception {
+        KeyManagerFactory keyManagerFactory = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+        keyManagerFactory.init(keyStore, "keystore_pass".toCharArray());
```

Comment:
"keystore_pass" is this password ignored? can't find where it is defined...

Reply:
Yeah, this one is ignored, the keystore was already initialized without one. It's never saved and it only holds public keys so this one isn't exactly sensitive.